### ISSUE TITLE
ImportControl: global blacklist

### ISSUE THREAD

**Issue by rnveach** [checkstyle/checkstyle#5726]
This may just be an issue with Checkstyle's own import control for now, but I noticed this when reviewing the import control file.

For all packages, we have [a list of banned imports of `java.io.File*Stream`, `java.util.stream.*`, and `java.util.Comparator`](https://github.com/checkstyle/checkstyle/blob/master/config/import-control.xml#L8-L24). However, the gui sub-package with it's own isolated list of imports, allows `java.io` and allows `java.util.*` (with no sub-subpackages).

This means `gui` allows some imports that are banned on a global level. [Because sub-packages are examined first](https://github.com/checkstyle/checkstyle/blob/master/src/xdocs/config_imports.xml#L892-L894), the only current way to correct this is to ban imports in 2 places, duplicating the list and reasoning.

Should we consider to create a new area that will be a global blacklist to ban imports everywhere and ignore other rules?
Or maybe split the import control into 2, a global blacklist and a package by package whitelist?

**Comment by romani** [comment]
I think it is problem/nuance of our config only.

We have `<subpackage name="gui" strategyOnMismatch="disallowed">`  if we change strategyOnMismatch to delegateToParent, it will work like you want.
This particular import ban planned to be addressed at #5616 .

**Comment by romani** [comment]
you are right.

Here is typical configuration problem, user is responsible to keep it reasonable. This nuance is by design :).  It is probably a good example of problems when both models (allow and disallow) are used.
I am more in favor of blacklist model. White list model also make sense some time. But mix of models might be problematic to support by user.

We can convert our import control file to blacklist model for all except for API. only in API use while-list model. Does it make sense ?

**Comment by rnveach** [comment]
> We can convert our import control file to 

If you want to do something like https://github.com/checkstyle/checkstyle/blob/master/src/xdocs/config_imports.xml#L1174-L1181 this is fine with me.

The only problem is making sure the disallowed list is complete and maybe this is why I prefer whitelist mode. With blacklist, you have to manually verify all imports are correct in new files and that no stray ones are accidentally being allowed in because you forgot a disallow. I say `manually` because there will be no new violation and no change in the import control file that specifies any of the new imports are truly 'ok' since it is a blanket allow. When we create new sub-packages or include new tools, we will have to manually re-examine and update the import control file.
With whitelist, there will be a violation always and the import control file must be modified when a new import is seen and this allows reviewers like us to confirm that the new import isn't breaking anything and is ok to proceed. The only time I have had issues with allows/disallows being intermingled was with https://github.com/checkstyle/checkstyle/issues/5642 .

> for all except for API. only in API ... use while-list model

This still leaves us with the issue of a global blacklist. Instead of with all sub-packages, you are just moving it to API sub-package. The sub-package will have allows that can override the main package's disallows. This still means we would have to duplicate the disallows and reasons to the sub-package to achieve a global blacklist.

**Comment by romani** [comment]
Can existing `local-only=false` help here ?

**Comment by rnveach** [comment]
If I understand what you are asking, `local-only=false` is the default option for all rules and doesn't make the rule global.
See https://github.com/checkstyle/checkstyle/blob/90f20e09869c71eb22190ad9c1d46d5deec324a5/src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportControlLoader.java#L146

All rules still follow the sub-package hierarchy first regardless of any attributes. `local-only`'s real purpose is just for setting it to `true` for now.

**Comment by rnveach** [comment]
> Does it sound like bug that this property does not do what it supposed to do ?
In my mind if local-only=false then rule should be applied to all subpackages

Not a bug and it does apply to all sub-packages, it is just not the first thing that is checked via the rules. 

ImportControl uses a bottom first approach (sub-packages before package). `local-only` still keeps the allow/disallow rule inside the package it is defined. The order of precedence still checks the rules of the subpackage first before the package. Nothing changes that sub-packages are looked at first.

allow `local-only=false` - a package or sub-package can use this import, but sub-package rules take precedence over this one
allow `local-only=true` - only a package can use this import, but a sub-package can still override it

We would need a whole new attribute that would copy allow/disallow imports into all the sub-packages and be examined first, hence a global list.
We can either use `global=true/false` or we can change `local-only` into a 3 value attribute.

With global on, it would turn:
````
<package="com.company">
  <disallow class="my.test" global="true"/>
  <subpackage="example">
  </subpackage>
  <subpackage="example2">
  </subpackage>
<package>
```` 
into:
````
<package="com.company">
  <disallow class="my.test"/>
  <subpackage="example">
    <disallow class="my.test"/>
  </subpackage>
  <subpackage="example2">
    <disallow class="my.test"/>
  </subpackage>
<package>
```` 
Basically turning this one disallow into a top first (package before sub-packages)

**Comment by romani** [comment]
Good points for documentation update.

Looks like we need some attribute  to let parents subpackages apply some rule to all children. New attribute `for-all-subpackages=true|false` false is default. Rules will be copied to all subpackages controls, it will be same as do copy manually.

Should it help to your use case ?

**Comment by rnveach** [comment]
> Should it help to your use case ?

Yes, that would create a global blacklist or allowlist.

**Comment by romani** [comment]
I recommend to close this issue (too much discussions in it ) and make two more:
To fix documentation for local-only
To introduce for-all-subpackages to support global blacklist or allowlist.
